aws/signer/v4/handler.go

```diff
@@ -0,0 +1,146 @@
+package v4
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws/client"
+	v4Internal "github.com/aws/aws-sdk-go-v2/aws/signer/internal/v4"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyHTTP "github.com/awslabs/smithy-go/transport/http"
+)
+
+const errPayloadNotSeekable = "request payload is not seekable"
+
+// HashComputationError indicates an error occurred while computing the signing hash
+type HashComputationError struct {
+	cause string
+	err   error
+}
+
+// Error is the error message
+func (e *HashComputationError) Error() string {
+	msg := "failed to compute payload hash"
+	if len(e.cause) == 0 {
+		return msg
+	} else {
+		return fmt.Sprintf("%s: %s", msg, e.cause)
+	}
+}
+
+// Unwrap returns the underlying error if one is set
+func (e *HashComputationError) Unwrap() error {
+	return e.err
+}
+
+// SigningError indicates an error condition occurred while performing SigV4 signing
+type SigningError struct {
+	cause string
+	err   error
+}
+
+func (e *SigningError) Error() string {
+	msg := "failed to sign request"
+	if len(e.cause) == 0 {
+		return msg
+	} else {
+		return fmt.Sprintf("%s: %s", msg, e.cause)
+	}
+}
+
+// Unwrap returns the underlying error cause
+func (e *SigningError) Unwrap() error {
+	return e.err
+}
+
+type payloadHashKey struct{}
+
+// UnsignedPayloadMiddleware sets the SigV4 request payload hash to unsigned
+type UnsignedPayloadMiddleware struct{}
+
+func (m *UnsignedPayloadMiddleware) Name() string {
+	return "SigV4 unsigned payload middleware"
+}
+
+func (m *UnsignedPayloadMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	ctx = context.WithValue(ctx, payloadHashKey{}, v4Internal.UnsignedPayload)
```

Comment:
What do you think about putting these WithValue, GetValue calls in helper utility functions? Their public interface would be fore the expected type of the payload.